packages/block-editor/src/hooks/generated-class-name.js

```diff
@@ -11,30 +22,41 @@ import { hasBlockSupport, getBlockDefaultClassName } from '@wordpress/blocks';
  *
  * @param {Object} extraProps Additional props applied to save element.
  * @param {Object} blockType  Block type.
+ * @param {Object} attributes Block attributes.
  *
  * @return {Object} Filtered props applied to save element.
  */
-export function addGeneratedClassName( extraProps, blockType ) {
-	// Adding the generated className.
-	if ( hasBlockSupport( blockType, 'className', true ) ) {
-		if ( typeof extraProps.className === 'string' ) {
-			// We have some extra classes and want to add the default classname
-			// We use uniq to prevent duplicate classnames.
-
-			extraProps.className = [
-				...new Set( [
-					getBlockDefaultClassName( blockType.name ),
-					...extraProps.className.split( ' ' ),
-				] ),
-			]
-				.join( ' ' )
-				.trim();
-		} else {
-			// There is no string in the className variable,
-			// so we just dump the default name in there.
-			extraProps.className = getBlockDefaultClassName( blockType.name );
+export function addGeneratedClassName( extraProps, blockType, attributes ) {
+	const generatedClassNames = [];
+	if ( hasBlockClassNameSupport( blockType ) ) {
+		generatedClassNames.push( getBlockDefaultClassName( blockType.name ) );
+	}
+	if ( hasVariationClassNameSupport( blockType ) ) {
+		const variationClassName = getBlockVariationClassName(
+			blockType.name,
+			attributes
+		);
+		if ( variationClassName ) {
+			generatedClassNames.push( variationClassName );
 		}
 	}
+	if ( typeof extraProps.className === 'string' ) {
```

Comment:
If we don't have either classname support at this point, we can bail early and return the props.